metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java

```diff
@@ -171,74 +172,125 @@ private void extractBroaderResources(EnrichmentBase resource, Set<String> destin
       }
+    } else {
+      // If the vocabulary previously used is not null then we accept that the previous
+      // transformation of the provider entity did not succeed, and we have cached that result
+      transformedEntityAndVocabularyPair = new ImmutablePair<>(transformedEntityXml,
+          entityVocabulary);
     }
 
-    // Parse the entity.
+    return transformedEntityAndVocabularyPair;
+  }
+
+  private Pair<EnrichmentBase, Vocabulary> convertToEnrichmentBaseVocabularyPair(String entityXml,
+      Vocabulary entityVocabulary) throws JAXBException {
     final Pair<EnrichmentBase, Vocabulary> result;
-    if (entityString == null || vocabulary == null) {
+    if (entityXml == null || entityVocabulary == null) {
       result = null;
     } else {
-      final StringReader reader = new StringReader(entityString);
+      final StringReader reader = new StringReader(entityXml);
       final JAXBContext context = JAXBContext.newInstance(EnrichmentBase.class);
-      final EnrichmentBase resource =
-          (EnrichmentBase) context.createUnmarshaller().unmarshal(reader);
-      result = new ImmutablePair<>(resource, vocabulary);
+      final EnrichmentBase resource = (EnrichmentBase) context.createUnmarshaller()
+          .unmarshal(reader);
+      result = new ImmutablePair<>(resource, entityVocabulary);
     }
-
     return result;
   }
 
-  private Pair<String, Vocabulary> retrieveTransformedEntity(String resourceId,
-      VocabularyCandidates candidates) throws TransformerException {
-
-    // Get the original entity given the list of vocabulary candidates
-    final String originalEntity = retrieveOriginalEntity(resourceId, candidates);
-
-    // If we could not resolve the entity, or there was no vocabulary, we are done.
-    if (originalEntity == null) {
-      return null;
-    }
+  private Pair<String, Vocabulary> retrieveEntityAndTransform(String resourceId,
+      VocabularyCandidates vocabularyCandidates) throws TransformerException {
 
-    // Transform the original entity.
-    for (Vocabulary candidate : candidates.getCandidates()) {
-      final String transformedEntity = transformEntity(candidate, originalEntity, resourceId);
-      if (transformedEntity != null) {
-        return new ImmutablePair<>(transformedEntity, candidate);
+    String originalEntity = retrieveOriginalEntity(resourceId, vocabularyCandidates);
+    //If original entity exists, try transformation
+    ImmutablePair<String, Vocabulary> transformedEntityAndVocabularyPair = null;
+    if (originalEntity != null) {
+      // Transform the original entity.
+      for (Vocabulary vocabulary : vocabularyCandidates.getVocabularies()) {
+        final String transformedEntity = retrieveEntityAndTransform(vocabulary, originalEntity,
+            resourceId);
+        if (transformedEntity != null) {
+          transformedEntityAndVocabularyPair = new ImmutablePair<>(transformedEntity, vocabulary);
+          break;
+        }
```

Comment:
I see that here (like you announced) if there is no vocabulary for a record, you return null and therefore it is not saved to the cache.
But is it not a good idea to save it to the cache regardless (with a null vocabulary)? If we can't find a vocabulary then we know that the next time we don't need to try. Could save some database access time.

Reply:
Well the functionality about the null vocabulary was replicated as before.
So a cache that would have a null value as vocabulary will be "retried", because at later stage that vocabulary would succeed. From what you are saying, we don't want that so we don't re-compute it and we just return null for that resourceId, right?
Also another reason was that a cache item can point to a vocabulary that does not exist anymore therefore it was null, but I can handle that in a different way..